File: backend/app/schemas/vfx/lineage.py
"""Lineage schemas for the technology genealogy graph."""
from pydantic import BaseModel, ConfigDict


class LineageNode(BaseModel):
    id: int  # Item ID
    title: str
    source: str
    priority: str | None = None
    llm_score: int = 0
    year: int | None = None
    url: str | None = None  # 이슈 #16 P2: manual item 은 URL 없음


class LineageEdgeRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int | None = None
    parent_id: int
    child_id: int
    relationship_type: str
    origin: str = "auto"       # auto | manual | arca
    status: str = "confirmed"  # confirmed | suggested
    note: str | None = None


class LineageEdgeCreate(BaseModel):
    parent_id: int
    child_id: int
    relationship_type: str = "related"
    note: str | None = None


class LineageGraph(BaseModel):
    center_id: int | None = None
    nodes: list[LineageNode]
    edges: list[LineageEdgeRead]

File: frontend/src/vfx/api/lineage.ts
import { apiGet, apiPost, apiDelete } from "./client";

export type LineageNode = {
  id: number;
  title: string;
  source: string;
  priority: string | null;
  llm_score: number;
  year: number | null;
  url: string | null;
};

export type LineageEdge = {
  id?: number;
  parent_id: number;
  child_id: number;
  relationship_type: string;
  origin?: string; // auto | manual | arca
  status?: string; // confirmed | suggested
  note?: string | null;
};

export type LineageGraph = {
  center_id: number | null;
  nodes: LineageNode[];
  edges: LineageEdge[];
};

export const fetchItemLineage = (itemId: number, depth = 2) =>
  apiGet<LineageGraph>(`/lineage/item/${itemId}?depth=${depth}`);

export const fetchCategoryLineage = (slug: string) =>
  apiGet<LineageGraph>(`/lineage/category/${slug}`);

// ── Phase 3 편집 (professor/admin) — 자유 엣지 + AI 추정 confirm ──
export const createLineageEdge = (body: {
  parent_id: number;
  child_id: number;
  relationship_type?: string;
  note?: string;
}) => apiPost<LineageEdge>(`/lineage/edges`, body);

export const confirmLineageEdge = (edgeId: number) =>
  apiPost<LineageEdge>(`/lineage/edges/${edgeId}/confirm`, {});

export const deleteLineageEdge = (edgeId: number) =>
  apiDelete<void>(`/lineage/edges/${edgeId}`);
